app/src/main/java/org/jellyfin/androidtv/data/eventhandling/SocketHandler.kt

```diff
@@ -0,0 +1,164 @@
+import org.jellyfin.apiclient.model.entities.MediaType
+import org.jellyfin.sdk.api.client.ApiClient
+import org.jellyfin.sdk.api.client.extensions.sessionApi
+import org.jellyfin.sdk.api.client.extensions.userLibraryApi
+import org.jellyfin.sdk.api.sockets.SocketInstance
+import org.jellyfin.sdk.api.sockets.addGeneralCommandsListener
+import org.jellyfin.sdk.api.sockets.addListener
+import org.jellyfin.sdk.model.api.GeneralCommandType
+import org.jellyfin.sdk.model.api.LibraryUpdateInfo
+import org.jellyfin.sdk.model.api.PlaystateCommand
+import org.jellyfin.sdk.model.extensions.getValue
+import org.jellyfin.sdk.model.serializer.toUUIDOrNull
+import org.jellyfin.sdk.model.socket.LibraryChangedMessage
+import org.jellyfin.sdk.model.socket.PlayMessage
+import org.jellyfin.sdk.model.socket.PlayStateMessage
+import timber.log.Timber
+import java.util.UUID
+
+class SocketHandler(
+	private val context: Context,
+	private val api: ApiClient,
+	private val dataRefreshService: DataRefreshService,
+	private val mediaManager: MediaManager,
+) {
+	private val coroutineScope = CoroutineScope(Dispatchers.IO)
+	private var socketInstance: SocketInstance? = null
+
+	suspend fun updateSession() {
+		api.sessionApi.postCapabilities(
+			playableMediaTypes = listOf(MediaType.Video, MediaType.Audio),
+			supportsMediaControl = true,
+			supportedCommands = listOf(
+				GeneralCommandType.DISPLAY_MESSAGE,
+				GeneralCommandType.SEND_STRING,
+			),
+		)
+
+		if (socketInstance != null) socketInstance?.updateCredentials()
+		else socketInstance = createInstance()
+	}
+
+	private fun createInstance() = api.ws().apply {
+		// Library
+		addListener<LibraryChangedMessage> { message -> onLibraryChanged(message.info) }
+
+		// Media playback
+		addListener<PlayMessage> { message -> onPlayMessage(message) }
+		addListener<PlayStateMessage> { message -> onPlayStateMessage(message) }
+
+		// General commands
+		addGeneralCommandsListener(setOf(GeneralCommandType.DISPLAY_CONTENT)) { message ->
+			val itemId by message
+			val itemUuid = itemId?.toUUIDOrNull()
+
+			if (itemUuid != null) onDisplayContent(itemUuid)
+		}
+		addGeneralCommandsListener(setOf(GeneralCommandType.DISPLAY_MESSAGE, GeneralCommandType.SEND_STRING)) { message ->
+			val header by message
+			val text by message
+			val string by message
+
+			onDisplayMessage(header, text ?: string)
+		}
+	}
+
+	private fun onLibraryChanged(info: LibraryUpdateInfo) {
+		Timber.d(buildString {
+			appendLine("Library changed.")
+			appendLine("Added ${info.itemsAdded!!.size} items")
+			appendLine("Removed ${info.itemsRemoved!!.size} items")
+			appendLine("Updated ${info.itemsUpdated!!.size} items")
+		})
+
+		if (info.itemsAdded!!.any() || info.itemsRemoved!!.any())
+			dataRefreshService.lastLibraryChange = System.currentTimeMillis()
+	}
+
+	private fun onPlayMessage(message: PlayMessage) {
+		val itemId = message.request.itemIds?.firstOrNull() ?: return
+
+		PlaybackHelper.retrieveAndPlay(
```

Comment:
With TvApiEventListener there was a guard against starting a video while one was already playing. I think it is better without it but it's possible there could be issues introduced.

I haven't seen any issues yet while testing but switching between videos like this relies on the overlay activity lifecycle to kill the prior session. So I think it's something to keep an eye on.

Reply:
If issues are found maybe they could be fixed in PlaybackHelper by checking to see if:
* the selected item is already playing - ignore the command in this case
* if an item is already playing, stop the player before starting the new item. This way there isn't any weird overlapping

Reply:
Yeah so I simplified this a lot because I too think the checks should not be in the class. While testing I didn't see any side effects from this changed behavior.